mcs/class/System/Mono.UnityTls/UnityTlsContext.cs

```diff
@@ -0,0 +1,370 @@
+			if (errorState.code == UnityTls.unitytls_error_code.UNITYTLS_USER_WOULD_BLOCK)
+				return false;
+
+			// Not done is not an error if we are server and don't ask for ClientCertificate
+			if (result == UnityTls.unitytls_x509verify_result.UNITYTLS_X509VERIFY_NOT_DONE && IsServer && !AskForClientCertificate)
+				Unity.Debug.CheckAndThrow (errorState, "Handshake failed", AlertDescription.HandshakeFailure);
+			else
+				Unity.Debug.CheckAndThrow (errorState, result, "Handshake failed", AlertDescription.HandshakeFailure);
+
+			// .Net implementation gives the server a verification callback (with null cert) even if AskForClientCertificate is false.
+			// We stick to this behavior here.
+			if (IsServer && !AskForClientCertificate) {
+				if (!ValidateCertificate (null, null))
+					throw new TlsException (AlertDescription.HandshakeFailure, "Verification failure during handshake");
+			}
+
+			return true;
+		}
+
+		public override void FinishHandshake ()
+		{
+			// Query some data. Ignore errors on the way since failure is not crucial.
+			var errorState = UnityTls.NativeInterface.unitytls_errorstate_create ();
+			var cipherSuite = UnityTls.NativeInterface.unitytls_tlsctx_get_ciphersuite(m_TlsContext, &errorState);
+			var protocolVersion = UnityTls.NativeInterface.unitytls_tlsctx_get_protocol(m_TlsContext, &errorState);
+
+			m_Connectioninfo = new MonoTlsConnectionInfo () {
+				CipherSuiteCode = (CipherSuiteCode)cipherSuite,
+				ProtocolVersion = UnityTlsConversions.ConvertProtocolVersion(protocolVersion),
+				PeerDomainName = ServerName
+
+				// TODO:
+				// The following properties can be deducted from CipherSuiteCode.
+				// It looks though like as of writing no Mono implemention fills it out and there is also no mechanism that does that automatically
+				//
+				//CipherAlgorithmType
+				//HashAlgorithmType
+				//ExchangeAlgorithmType
+			};
+			m_IsAuthenticated = true;
+		}
+
+		[MonoPInvokeCallback (typeof (UnityTls.unitytls_tlsctx_write_callback))]
+		static private size_t WriteCallback (void* userData, byte* data, size_t bufferLen, UnityTls.unitytls_errorstate* errorState)
+		{
+			var handle = (GCHandle)(IntPtr)userData;
+			var context = (UnityTlsContext)handle.Target;
+			return context.WriteCallback (data, bufferLen, errorState);
+		}
+
+		private size_t WriteCallback (byte* data, size_t bufferLen, UnityTls.unitytls_errorstate* errorState)
+		{
+			try {
+				if (m_WriteBuffer == null || m_WriteBuffer.Length < bufferLen)
+					m_WriteBuffer = new byte[bufferLen];
+				Marshal.Copy ((IntPtr)data, m_WriteBuffer, 0, bufferLen);
+
+				if (!Parent.InternalWrite (m_WriteBuffer, 0, bufferLen)) {
+					UnityTls.NativeInterface.unitytls_errorstate_raise_error (errorState, UnityTls.unitytls_error_code.UNITYTLS_USER_WRITE_FAILED);
+					return 0;
+				}
+
+				return bufferLen;
+			} catch { // handle all exceptions since we don't want to let them go through native code.
```

Comment:
I wonder if this is really okay... I imagine this is a fairly common error. Like dropping a network connection or some other io failure. In which case we will hide the real reason for the failure.

Reply:
Don't think it is great either. But if we jump out via exception here, the unitytls impl ends up in an invalid state. I'd call that worse. I guess we could maybe always "cache" the last exception and throw it when we're outside of the read/write/verify block (whenever cached exception is non zero and we registered an error in handshake/read/write). What do you think?

Reply:
How do the other TLS implementations handle this?

Reply:
Turns out AppleTls does exactly what I described. So I better get on with it :)

Reply:
Done. Also added tests for user exception forwarding

Reply:
done